src/plugins/interactive_setup/server/verification_code.ts

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import chalk from 'chalk';
+import crypto from 'crypto';
+
+import type { Logger } from 'src/core/server';
+
+import { VERIFICATION_CODE_LENGTH } from '../common';
+
+export class VerificationCode {
+  public readonly code = VerificationCode.generate(VERIFICATION_CODE_LENGTH);
+  private failedAttempts = 0;
+  private readonly maxFailedAttempts = 5;
+
+  constructor(private readonly logger: Logger) {}
+
+  public get remainingAttempts() {
+    return this.maxFailedAttempts - this.failedAttempts;
+  }
+
+  public verify(code: string | undefined) {
+    if (this.failedAttempts >= this.maxFailedAttempts) {
+      this.logger.error(
+        'Maximum number of attempts exceeded. Restart Kibana to generate a new code and retry.'
```

Comment:
__suggestion:__ it's be great if we suggest to restart in the UI too - from there it's not clear what to do. 

__suggestion:__ Also if all attempts failed and I restart, Kibana only displays a new code as a part of the URL, but I'm in the process of manual configuration already, entered all the data and now just need a code, I don't want to use a link and start over - would be nice if we also print code right after we print URL with it.

Reply:
> **suggestion:** it'd be great if we suggest to restart in the UI too - from there it's not clear what to do.

Good point, added.

> **suggestion:** Also if all attempts failed and I restart, Kibana only displays a new code as a part of the URL, but I'm in the process of manual configuration already, entered all the data and now just need a code, I don't want to use a link and start over - would be nice if we also print code right after we print URL with it.

We can't detect whether you're the middle of the enrollment flow after restarting Kibana and I don't want to output unnecessary information and confuse users with codes they don't need to know. The assumption is that users have or need the enrollment token in their clipboard. I don't want them to override that with the verification code unless they absolutely need that. 

I hope most users will click the link to get to the enrollment flow and will never need to enter the verification token let alone restart kibana so I wouldn't optimise for this case.